src/libexpr/primops/fromYAML.cc

```diff
@@ -0,0 +1,156 @@
+
+            for (ryml::ConstNodeRef child : t.children()) {
+                std::string_view key(child.key().begin(), child.key().size());
+                visitYAMLNode(context, attrs.alloc(key), child);
+            }
+
+            v.mkAttrs(attrs);
+        }
+    } else if (t.is_seq()) {
+        context.state.mkList(v, t.num_children());
+
+        size_t i = 0;
+        for (ryml::ConstNodeRef child : t.children())
+            visitYAMLNode(context, *(v.listElems()[i++] = context.state.allocValue()), child);
+    } else if (valTypeCheck(ryml::TAG_NULL) && t.val_is_null()) {
+        v.mkNull();
+    } else if (t.has_val()) {
+        bool _bool;
+        NixFloat _float;
+        NixInt _int;
+        bool isQuoted = t.is_val_quoted();
+        auto val = t.val();
+        // Caution: ryml is able to convert integers into booleans and ryml::from_chars might ignore trailing chars
+        if (t.has_val_tag() && t.val_tag() == "!" && val.empty() && !isQuoted) { // special case for YAML string "!"
+            v.mkNull();
+        } else if (valTypeCheck(ryml::TAG_INT, !isQuoted) && val.is_integer() && ryml::from_chars(val, &_int)) {
+            v.mkInt(_int);
+        } else if (valTypeCheck(ryml::TAG_BOOL, !isQuoted) && ryml::from_chars(val, &_bool)) {
+            v.mkBool(_bool);
+        } else if (valTypeCheck(ryml::TAG_FLOAT, !isQuoted) && val.is_number() && ryml::from_chars(val, &_float)) {
+            v.mkFloat(_float);
+        }
+        if (valTypeCheck(ryml::TAG_STR) && v.type() == nThunk) {
+            std::string_view value(val.begin(), val.size());
+            v.mkString(value);
+        }
+    }
+    if (v.type() == nThunk) {
+        auto val = t.has_val() ? t.val() : "";
+        auto tag = t.has_val_tag() ? t.val_tag() : "";
+        auto msg = ryml::formatrs<std::string>(
+            "Error: The YAML value '{}' with '{}' tag cannot be represented as Nix data type", val, tag);
+        s_error(msg.data(), msg.size(), {}, &context);
+    }
+}
+
+static RegisterPrimOp primop_fromYAML({
+    .name = "__fromYAML",
+    .args = {"e"},
+    .doc = R"(
+      Convert a YAML 1.2 string to a Nix value, if a conversion is possible. For example,
+
+      ```nix
+      builtins.fromYAML ''{x: [1, 2, 3], y: !!str null, z: null}''
+      ```
+
+      returns the value `{ x = [ 1 2 3 ]; y = "null"; z = null; }`.
+
+      Maps are converted to attribute sets, but attribute sets require String keys, so that no other key data types are supported.
+
+      Scalars are converted to the type specified by their optional value tag and parsing fails, if a conversion is not possible.
+      Not all YAML types are supported by Nix, e.g. Nix has no binary and timestamp data types, so that parsing of YAML with any of these types fails.
+      Custom tags are ignored and a stream with multiple documents is mapped to a list except when the stream contains a single document.
```

Comment:
Should custom tags be ignored or should they throw an error? Or maybe map to the "type" attribute?

Reply:
I think throwing an error might be a good starting point, assuming it would not be a breaking change to turning an error into a non-error in the future.

Reply:
Timestamps are parsed as strings at the moment. YAML 1.1 has base60 encoded numbers (no longer mentioned in YAML 1.2), which don't have any prefix, but have `:` at specific positions. This is very questionable and it's not supported by rapidyaml anyhow.
In my opinion it makes no sense to parse timestamps, unless the format is well specified. And even this would be difficult without explicit tag. For example `1.23` is parsed as float. With correct rounding this can still be converted to the string `"1.23"`. Of course this property breaks in case of `1.230`. If something would be parsed as timestamp, then it is probably impossible to recover the original representation, unless the representation would be saved. But then equivalent values can have different representations...
I propose to support JSON schema tags only. If another tag is encountered, then an error can be thrown. I'm not sure whether lazy errors are a good solution. The tree generation is not lazy and thus on completion I expect that nix was able to parse the input. Throwing an error immediately is more restrictive than lazy errors. Thus lazy errors could be introduced later, too.